### PR TITLE
exporter: Added https(TLSv13) support

### PR DESCRIPTION
**Pull request by jmolmo**
Ceph exporter can run now as http or https server

Added two new parameters in order to provide the certificate files needed to use https.

```
jolmomar:~/.../ceph/build$ ./bin/ceph-exporter -h
usage: ceph-exporter [options]
options:
  ...
  --cert-file:    Path to the certificate file to use for https
  --key-file:     Path to the certificate key file to use for https
  ....
```
  
**Key facts:**
- If no certificates files provided using the parameters, the exporter is launched as http server. Exactly the same behavior we have until now. This assures us complete compatibility with the previous exporter version and therefore no impact at all in other software pieces using/configuring Ceph exporter.
- If the certificate and key file are provided and they are valid , the Ceph exporter starts as https server.
- Any error/omission in the certificate parameters or inside the certificate files will cause that the ceph exporter does not start. This is intentional, because in these cases, to fallback to http can drive to further problems thinking that the server is running https.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/exporter/ceph_exporter.cc [hunk 1]
```diff
@@ -1,7 +1,7 @@
 #include "common/ceph_argparse.h"
 #include "common/config.h"
 #include "exporter/DaemonMetricCollector.h"
-#include "exporter/http_server.h"
+#include "exporter/web_server.h"
```
Comment (epuertat) [inline src/exporter/ceph_exporter.cc:4]: I'm curious, why this change? HTTP is still a valid name even if you use HTTP + TLS. TLS is actually at the transport layer.

IMHO a "webserver" serves HTML formatted documents (plus CSS, Javascript and other assets), while Prometheus endpoints are just plain text HTTP servers.

File: src/exporter/web_server.cc [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+    if (request_.target() == "/") {
+      response_.set(http::field::content_type, "text/html; charset=utf-8");
+      std::string body("<html>\n"
+                        "<head><title>Ceph Exporter</title></head>\n"
+                        "<body>\n"
+                        "<h1>Ceph Exporter</h1>\n"
+                        "<p><a href='/metrics'>Metrics</a></p>"
+                        "</body>\n"
+                        "</html>\n");
+      response_.body() = body;
```
Comment (epuertat) [inline src/exporter/web_server.cc:78]: Instead of providing a formatted HTML document linking to the /metrics endpoints, I'd rather return a plain-text or even empty doc with a 301 (Moved permanently) HTTP status.

File: src/exporter/web_server.cc [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+
+  http_server(acceptor, socket);
+
+  dout(1) << "HTTP server running on " << exporter_addr << ":" << port << dendl;
+  ioc.run();
+}
+
+void run_https_server(const std::string& exporter_addr, unsigned short port, const std::string& cert_file, const std::string& key_file) {
+  net::io_context ioc{1};
+  ssl::context ssl_ctx(ssl::context::tlsv12);
```
Comment (epuertat) [inline src/exporter/web_server.cc:250]: Why not 1.3? Some cyphers in 1.2 are known to be vulnerable.

File: src/exporter/web_server.cc [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+  https_server(acceptor, ssl_ctx);
+
+  dout(1) << "HTTPS server running on " << exporter_addr << ":" << port << dendl;
+  ioc.run();
+}
+
+void web_server_thread_entrypoint() {
+  try {
+    std::string exporter_addr = g_conf().get_val<std::string>("exporter_addr");
+    unsigned short port = g_conf().get_val<int64_t>("exporter_http_port");
```
Comment (epuertat) [inline src/exporter/web_server.cc:265]: TCP ports can't have more than 16 bits (unsigned)

**Review by epuertat — COMMENTED** [review]
Not familiar with the Beast library, so I left some comments on security and usability.

File: src/exporter/web_server.cc (hunk 2 above)
Comment (jmolmo) [reply src/exporter/web_server.cc:78]: mmm... that it was exactly what we have previously. 
if we change that we can introduce some unexpected regression problem .... (very improbable, but possible ... who know what users have done interacting with ceph exporter)
Again.. I only have this concern about regression. but if nobodey shares my worries, nothing against and i will do the change.

Comment (rkachach) [reply src/exporter/web_server.cc:78]: AFAIK from discussions with other colleagues nobody uses `ceph-exporter` when ceph is deployed in external mode mainly because it was not considered safe.

File: src/exporter/web_server.cc (hunk 4 above)
Comment (jmolmo) [reply src/exporter/web_server.cc:265]: changed

File: src/exporter/ceph_exporter.cc (hunk 1 above)
Comment (jmolmo) [reply src/exporter/ceph_exporter.cc:4]: I leave it as [web server](https://en.wikipedia.org/wiki/Web_server)

File: src/exporter/web_server.cc (hunk 2 above)
Comment (jmolmo) [reply src/exporter/web_server.cc:78]: Changed: 


Now... taking a look to the response :
```
# curl -v -k https://localhost:9926/
....
< HTTP/1.1 301 Moved Permanently
< Location: /metrics
< Content-Length: 0
< 
* Connection #0 to host localhost left intact
# 
```

apparently, no answer if we do not take a loot to the response fields:
```
# curl -k https://localhost:9926/
#
```

File: src/common/options/ceph-exporter.yaml.in [hunk 5]
```diff
@@ -25,6 +25,20 @@ options:
   default: 9926
   services:
   - ceph-exporter
+- name: exporter_cert_file
+  type: str
+  level: advanced
```
[the remark this answers is not in the document]
Comment (jmolmo) [reply src/common/options/ceph-exporter.yaml.in:30]: Only found certificates management in RGW, and they are using "advanced"

File: src/exporter/web_server.cc (hunk 4 above)
Comment (jmolmo) [reply src/exporter/web_server.cc]: reverted.... There is a compilation warning if we use just a "short" type:

```
web_server.cc: In function ‘void run_http_server(const std::string&, short int)’:
    /home/jolmomar/Code/ceph/src/exporter/web_server.cc:232:70: warning: narrowing conversion of ‘port’ from ‘short int’ to ‘boost::asio::ip::port_type’ {aka ‘short unsigned int’} [-Wnarrowing]
      232 |   tcp::acceptor acceptor{ioc, {net::ip::make_address(exporter_addr), port}};
```

File: src/exporter/web_server.cc (hunk 2 above)
Comment (avanthakkar) [reply src/exporter/web_server.cc]: Yes! Let's keep it as it used to be, I'd not personally like to change any patterns here & given we already have been following prometheus exporter, we should resemble the same here & expose html page

**Review by avanthakkar — APPROVED** [review]
LGTM functionally!

Screenshot from 2024-07-09 12-00-01

File: src/exporter/ceph_exporter.cc (hunk 1 above)
Comment (rkachach) [reply src/exporter/ceph_exporter.cc:4]: @jmolmo even if we finally decide with going with the name change I'm wondering why git is not able to track the name change here. It should be possible to see the changes yo have introduce but it seems like we are deleting a file and adding a new one. Is that the case? did you rename the file first before applying the patch?

**Review by epuertat — APPROVED** [review]

File: src/exporter/ceph_exporter.cc (hunk 1 above)
Comment (jmolmo) [reply src/exporter/ceph_exporter.cc:4]: @rkachach. yes. I deleted the old file, not renamed. 
(I think probably because becauseI worked with web_server.cc without deleting the http_server.cc file ). And when I decided to put the pull request ,I removed all the old and not used files.

Comment (epuertat) [reply src/exporter/ceph_exporter.cc:4]: Git usually is smart enough to track deletions + additions as file renames... In this case, the filename change might complicate tracking changes. @jmolmo , is there any practical reason to rename it?

Comment (jmolmo) [reply src/exporter/ceph_exporter.cc:4]: My fault @epuertat . It would be more easy the reviews if i did use the "mv" ... Taking into account time constraints, resources available (time), that the code is tested , and easy to "rebuild" the previous version. i am going to merge.
Thanks for pointing this .. next time i will be more careful!

**Comment by rkachach** [comment]
jenkins test make check

**Comment by rkachach** [comment]
jenkins test api

**Comment by rkachach** [comment]
jenkins test make check

**Comment by rkachach** [comment]
jenkins test api

**Comment by rkachach** [comment]
jenkins test api
